File: DataStructsAndAlgos/implementations/javascript/src/DataStructures/Trees/Heap/Heap.js
import { TreeNode } from "../TreeNode";

export class MinHeap {}

export class ArrayMinHeap {
  constructor() {
    this.heap = [];
  }
  left(parentElement) {
    return 2 * parentElement + 1;
  }
  right(parentElement) {
    return 2 * parentElement + 2;
  }
  getParent(childElement) {
    if (childElement % 2 === 0) return (childElement - 2) / 2; //child is right
    return (childElement - 1) / 2; //child is left
  }
  isNull(index) {
    return index >= this.length() || index < 0;
  }
  swap(a, b) {
    let temp = this.heap[a];
    this.heap[a] = this.heap[b];
    this.heap[b] = temp;
  }
  trickleUp(childIndex, parentIndex) {
    while (this.heap[parentIndex] > this.heap[childIndex]) {
      this.swap(childIndex, parentIndex);
      childIndex = parentIndex;
      parentIndex = this.getParent(parentIndex);
    }
  }
  heapify(root) {
    //Check children
    let rootRight = this.right(root);
    let rootLeft = this.left(root);
    while (
      (!this.isNull(rootRight) && this.heap[root] > this.heap[rootRight]) ||
      (!this.isNull(rootLeft) && this.heap[root] > this.heap[rootLeft])
    ) {
      let newChild;
      if (
        (!this.isNull(rootLeft) && this.isNull(rootRight)) ||
        this.heap[rootRight] > this.heap[rootLeft]
      )
        newChild = rootLeft;
      else newChild = rootRight;
      this.swap(root, newChild);
      root = newChild;
      rootRight = this.right(root);
      rootLeft = this.left(root);
    }
  }
  sort() {
    let sorted = [];
    //n
    while (this.length() > 0) {
      sorted.push(this.remove()); //logn
    }
    return sorted;
  }
  push(newElem) {
    this.heap.push(newElem);
    if (this.length() > 1) {
      let newElementIndex = this.length() - 1;
      this.trickleUp(newElementIndex, this.getParent(newElementIndex));
    }
  }
  remove() {
    let removed = this.heap[0];
    let endElement = this.heap.pop();
    if (this.length() > 0) {
      this.heap[0] = endElement;
      this.heapify(0);
    }
    return removed;
  }
  length() {
    return this.heap.length;
  }
  print() {
    return this.heap;
  }
}

File: DataStructsAndAlgos/implementations/python/src/ds/Trees/heap.py
class Heap:
    def __init__(self, elem):
        self.heap = []
        self.heap.append(elem)

    def getRightChildIndex(self, parent):
        return 2 * parent + 2

    def getLeftChildIndex(self, parent):
        return 2 * parent + 1

    def getParent(self, child):
        if child % 2 == 0:
            return int((child-2) / 2)
        else:
            return int((child-1) / 2)

    def swap(self, a, b):
        temp = self.heap[a]
        self.heap[a] = self.heap[b]
        self.heap[b] = temp

    def isNull(self, index):
        return index < 0 or index > len(self.heap)-1

    def sort(self):
        sortedHeap = []
        while(self.getLength() > 0):
            sortedHeap.append(self.remove())
        return sortedHeap

    def insert(self, newElem):
        self.heap.append(newElem)
        self.trickleUp()

    def remove(self):
        removed = self.heap[0]
        self.heap[0] = self.heap[len(self.heap)-1]
        self.heap.pop(len(self.heap)-1)
        self.heapify(0)
        return removed

    def heapify(self, root):
        rightChild = self.getRightChildIndex(root)
        leftChild = self.getLeftChildIndex(root)
        while not self.isNull(rightChild) and self.heap[rightChild] < self.heap[root] or not self.isNull(leftChild) and self.heap[leftChild] < self.heap[root]:
            newChild = -1
            if self.isNull(rightChild) and not self.isNull(leftChild) or self.heap[rightChild] > self.heap[leftChild]:
                newChild = leftChild
            else:
                newChild = rightChild
            self.swap(root, newChild)
            root = newChild
            rightChild = self.getRightChildIndex(root)
            leftChild = self.getLeftChildIndex(root)

    def trickleUp(self):
        currIndex = len(self.heap)-1
        parentIndex = self.getParent(currIndex)
        while not self.isNull(parentIndex) and self.heap[currIndex] < self.heap[parentIndex]:
            self.swap(currIndex, parentIndex)
            currIndex = parentIndex
            parentIndex = self.getParent(parentIndex)

    def getLength(self):
        return len(self.heap)

    def printHeap(self):
        return self.heap
